### PR TITLE
TestServer: mark response headers IsReadOnly on start

### PR DESCRIPTION
**Pull request by Tratcher**
#1253 This makes TestServer act more like a real server and will help us find potential product bugs.

### CODE REVIEW

**Comment by Tratcher** [comment]
I understand now why we're getting some flaky tests after the TestServer refactor.  The difference is that ClientHandler used to copy all of the response fields before it asynchronously diverged from the server thread.
https://github.com/aspnet/Hosting/pull/1248/files#diff-22356e5b7de1c9d8b3046bd1c6418043L207
Now HttpContextBuilder controls the threading and asynchronously diverges from the server thread before ClientHandler copies the response fields. The first change in this PR for IsReadOnly is still appropriate and will help us track down what test or component is incorrectly modifying response headers after the response starts. I've made an additional change to copy the feature collection so we can read it without hitting multi-threading issues like the following:
```
System.NullReferenceException : Object reference not set to an instance of an object.
   at Microsoft.AspNetCore.Http.Internal.DefaultHttpResponse.get_StatusCode()
   at Microsoft.AspNetCore.TestHost.ClientHandler.<SendAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.ResponseCaching.Tests.ResponseCachingTests.<ServesFreshContent_If_PragmaNoCache>d__6.MoveNext() in C:\BuildAgent\work\33bdfc1cae7b2a38\modules\ResponseCaching\test\Microsoft.AspNetCore.ResponseCaching.Tests\ResponseCachingTests.cs:line 164
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```
http://aspnetci/viewLog.html?buildId=353102&tab=buildResultsDiv&buildTypeId=XPlat_Windows_Win10Universe

File: src/Microsoft.AspNetCore.TestHost/HttpContextBuilder.cs [hunk 1]
```diff
@@ -110,7 +110,13 @@ internal void AbortRequest()
             {
                 // Sets HasStarted
                 await _responseFeature.FireOnSendingHeadersAsync();
-                _responseTcs.TrySetResult(_httpContext);
+                // Copy the feature collection so we're not multi-threading on the same collection.
+                var newFeatures = new FeatureCollection();
+                foreach (var pair in _httpContext.Features)
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.TestHost/HttpContextBuilder.cs]: Almost wish there is a FeatureCollection.Clone() method.

File: src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs [hunk 2]
```diff
@@ -13,10 +13,11 @@ internal class ResponseFeature : IHttpResponseFeature
     {
         private Func<Task> _responseStartingAsync = () => Task.FromResult(true);
         private Func<Task> _responseCompletedAsync = () => Task.FromResult(true);
+        private HeaderDictionary _headers = new HeaderDictionary();
 
         public ResponseFeature()
         {
-            Headers = new HeaderDictionary();
+            Headers = _headers;
             Body = new MemoryStream();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs]: Note to self, StatusCode and ReasonPhrase should also check HasStarted and throw

File: src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs [hunk 3]
```diff
@@ -13,20 +13,51 @@ internal class ResponseFeature : IHttpResponseFeature
     {
         private Func<Task> _responseStartingAsync = () => Task.FromResult(true);
         private Func<Task> _responseCompletedAsync = () => Task.FromResult(true);
+        private HeaderDictionary _headers = new HeaderDictionary();
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs]: Why bother with the explicit backing field for `Headers`? Don't need the concrete type anywhere.

File: test/Microsoft.AspNetCore.TestHost.Tests/ResponseFeatureTests.cs [hunk 4]
```diff
@@ -40,5 +41,28 @@ public void OnStarting_ThrowsWhenHasStarted()
                 }, state: "string");
             });
         }
+
+        [Fact]
+        public void StatusCode_ThrowsWhenHasStarted()
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.TestHost.Tests/ResponseFeatureTests.cs]: Suggest adding a similar test for `ReasonPhrase`.

File: src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs (hunk 3 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs]: Yes you do. IHeaderDictionary.IsReadOnly does not have a setter, only HeaderDictionary.IsReadOnly adds a setter.

File: test/Microsoft.AspNetCore.TestHost.Tests/ResponseFeatureTests.cs (hunk 4 above)
Comment (Tratcher) [reply test/Microsoft.AspNetCore.TestHost.Tests/ResponseFeatureTests.cs]: I covered reason phrase in this test

File: src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs (hunk 3 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs]: thx

**Pull request merged**